### ISSUE TITLE
EOFException when querying for a document with byte[]

### ISSUE THREAD

**Issue by criedel** [vznet/mongo-jackson-mapper#6]
I'm having a collection with lots of binary data in it. When looking for existing documents the query I wrote fails for some documents. And when it fails it always fails for the same document. I wasn't really able to track down the exact cause so I wrote a test case that adds one file a lot of times to a random collection and performs a query that is pretty identical to the one I use in my code.
It's kind of brute force but it seems to work (currently a few times in a row, so I'm confident you can see the same error).
When I'm debugging the line that throws the EOFException in LittleEndianInputStream.java I can reproduce the error every time. 

The test case produces this output on my local computer and my build server:
(unfortunately I committed it under #5)

```
Running net.vz.mongodb.jackson.TestParsingAndGenerating
Jan 1, 2012 11:16:19 PM com.mongodb.DBPortPool gotError
WARNING: emptying DBPortPool to 127.0.0.1:27017 b/c of error
java.io.EOFException
    at de.undercouch.bson4jackson.io.LittleEndianInputStream.readFully(LittleEndianInputStream.java:116)
    at de.undercouch.bson4jackson.io.LittleEndianInputStream.readFully(LittleEndianInputStream.java:108)
    at de.undercouch.bson4jackson.BsonParser.handleBinary(BsonParser.java:388)
    at de.undercouch.bson4jackson.BsonParser.nextToken(BsonParser.java:241)
    at org.codehaus.jackson.map.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:693)
    at org.codehaus.jackson.map.deser.BeanDeserializer.deserialize(BeanDeserializer.java:580)
    at org.codehaus.jackson.map.ObjectMapper._readValue(ObjectMapper.java:2695)
    at org.codehaus.jackson.map.ObjectMapper.readValue(ObjectMapper.java:1308)
    at net.vz.mongodb.jackson.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:62)
    at com.mongodb.Response.<init>(Response.java:66)
    at com.mongodb.DBPort.go(DBPort.java:128)
    at com.mongodb.DBPort.call(DBPort.java:79)
    at com.mongodb.DBTCPConnector.call(DBTCPConnector.java:218)
    at com.mongodb.DBApiLayer$MyCollection.__find(DBApiLayer.java:305)
    at com.mongodb.DBCursor._check(DBCursor.java:369)
    at com.mongodb.DBCursor._hasNext(DBCursor.java:498)
    at com.mongodb.DBCursor.hasNext(DBCursor.java:523)
    at net.vz.mongodb.jackson.DBCursor.hasNext(DBCursor.java:315)
    at net.vz.mongodb.jackson.TestParsingAndGenerating.testHugeData(TestParsingAndGenerating.java:313)
```

**Comment by jroper** [comment]
Ok, guessing this is either a bug in bson4jackson or in MongoDB.  I will take a further look into it, but could you answer the following questions:

How many bytes are in the binary data you are testing?
What version of MongoDB are you running?

**Comment by criedel** [comment]
I'd rather say it's bson4jackson or the mapper than mongodb since I can query for the data without problems when I use the driver API instead of DBQuery and the wrapped JacksonDBCollection:

```
            DBObject ref = new BasicDBObject("id", "4edf9b975e73d65678000762");
            ref.put("name", "ANAME");
            ref.put("size", "ASIZE");
            final com.mongodb.DBCursor find = movieImageDAO.db().getCollection("MyCollection").find(ref);

            if (find.hasNext()) {
                final byte[] data = (byte[]) find.next().get("data"); // always works
            }
```

The amount of bytes does not seem to be the cause. It happens with the "dataFile" from my latest commit. But only sometimes. My app is running for a while, binary data gets stored, documents created (or updated) and then there's this error. Totally random. I.e. if I clear the db and run the same procedure again it will fail with other documents and other byte[]. But it always fails after some time and it fails for a lot more documents, too. That's pretty hard to unit test, I know :-/ 

The code fails on mongodb 2.0.1 and 2.0.2.

**Comment by criedel** [comment]
Just had a look at my server logs and it seems it also breaks when parsing non-byte[] fields:

```
Caused by: java.io.EOFException
        at de.undercouch.bson4jackson.io.LittleEndianInputStream.readUnsignedByte(LittleEndianInputStream.java:151)
        at de.undercouch.bson4jackson.io.LittleEndianInputStream.readByte(LittleEndianInputStream.java:144)
        at de.undercouch.bson4jackson.io.LittleEndianInputStream.readUTF(LittleEndianInputStream.java:292)
        at de.undercouch.bson4jackson.io.LittleEndianInputStream.readUTF(LittleEndianInputStream.java:264)
        at de.undercouch.bson4jackson.BsonParser.readCString(BsonParser.java:481)
        at de.undercouch.bson4jackson.BsonParser.nextToken(BsonParser.java:212)
        at org.codehaus.jackson.map.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:690)
        at org.codehaus.jackson.map.deser.BeanDeserializer.deserialize(BeanDeserializer.java:580)
        at org.codehaus.jackson.map.ObjectMapper._readValue(ObjectMapper.java:2695)
        at org.codehaus.jackson.map.ObjectMapper.readValue(ObjectMapper.java:1308)
        at net.vz.mongodb.jackson.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:62)
```

Or is it just some weird load issue?

**Comment by pmaedel** [comment]
comparing LittleEndianInputStream.readFully() of bson4jackons to ByteStreams.readFully() of google guava makes me wonder why EOF is raised as within the former at all. Within the latter, an EOF is silently ignored when reading the stream which also seems to make more sense.

same is true for readUnsignedByte()

**Comment by jroper** [comment]
Thanks for your help with this, I've found the bug, it's actually in code that I contributed to bson4jackson, in the BoundedInputStream, which is returning the length of bytes request to be read on read(), rather than the actual number of bytes read.  Most times the OS buffers are able to supply these bytes, but very occasionally they can't, which is why this issue happens so infrequently.  I'll get the fix pushed into bson4jackson, and hopefully we can get a bson4jackson release soon.  A simple work around is to not use bson4jackson, by calling JacksonDBCollection.disable(JacksonDBCollection.Feature.USE_STREAM_DESERIALISATION).  The performance hit of not using bson4jackson here is not much.
